### ISSUE TITLE
Custom updater for Windows

### ISSUE THREAD

**Issue by allexzander** [nextcloud/desktop#6526]
We need to implement the updater similar to what is done in e.g. Chrome, or Sparkle on mac, such that we won't need to run the MSI but will just download what has changed (DLLs) and replace the old one with new ones during the restart. It is a big feature. The logic would also unload/kill the DLL extensions if needed (CF API shell extension and those for status overlays and a context menu). In most cases, we won't even need to kill them as they rarely change, but with the current approach - MSI - they are getting removed and re-installed on each update which is not ideal.

Even better would be to have a separate app that will get distributed and what it will do is download the binaries for the Nextcloud client and place them into the installation folder. This would be the second part of the feature and will make a big impact, so we will need to decide how to do it right and transition smoothly.

### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you want to have the same feature implemented.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments. 

### Feature description
Tell us how the feature should work

**Comment by allexzander** [comment]
@tobiasKaminsky @mgallien

**Comment by roberix** [comment]
Hi. 

@allexzander 

As much as I hate the current update process as enterprise customer (we do relulary have severe issues during updating) it gives me also the control I need. What we do is deploying the MSI (with autoupdate OFF) and we are fully in charge which version is installed.

Nextcloud is for our company a critial asset and judging the client versions which are sometimes released we do need to maintain full control. Giving away this control to the users (which will hump on every update button you will present to them) is a dangerous path.

Please whatever you implement - please consire to give the enterprise admins control over the client version management somehow.

On a securty side note - what you suggest makes a devastating supply chain attack more severe and more likely. If there will be a compromise then a self update can be a fatal event. We saw such events regulary in the recent past (Solarwinds..)

Best regards

rob

**Comment by allexzander** [comment]
@roberix Yes, this is something that I expected. So this needs to be designed thoroughly first.

**Comment by PhilippSchlesinger** [comment]
@allexzander 
It might be helpful to compare the requirements with the solution that Mozilla uses to update Firefox and Thunderbird, 
[Mozilla Maintenance Service](https://support.mozilla.org/en-US/kb/what-mozilla-maintenance-service)

At least some of the features needed for Nextcloud Client like installation without user intervention, installation without admin priviledges are solved there already.

**Comment by mgallien** [comment]
there would still a requirements to kill all processes that have loaded our files explorer extension DLLs
that is a significant difference with what firefox needs to do

**Comment by PhilippSchlesinger** [comment]
> there would still a requirements to kill all processes that have loaded our files explorer extension DLLs that is a significant difference with what firefox needs to do

I just wanted to point to a solution that manages installation of updates without user intervention and independently from the running software.

There are so many comments (https://github.com/nextcloud/desktop/issues/5369#issuecomment-1930885425 being only one of them) that indicate that not restarting explorer and not restarting Windows after installation works just fine.
Sure, that way running shell extensions will stay active and the updated shell extensions will get active only at the next start.

[Here](https://stackoverflow.com/questions/71158991/wix-how-to-restart-windows-explorer-to-install-uninstall-a-shell-extension) is a description of how exactly this workflow has been realised in an installer.

**Comment by jimklimov** [comment]
Can you make a shim DLL that Explorer et al can load, which is so little and stupid that it rarely if ever needs an update, and its job is to load/unload the actual DLL during update and proxy its resources?

Waaaay back in the day I remember the mail client TheBat! (IIRC) change its localization this way, by hot-reloading a different resource DLL...

**Comment by beedell-roke** [comment]
Shouldn't this issue be renamed to "delta updates for Windows", since *that*'s the desire? How it's accomplished - whether via a custom installer or not - is an implementation detail.
